In [1]:
import torch as t

In order to get familiar with the concept of a compytation graph, we will create one for the following function

<div class="MathJax_Display" style="text-align: center;">

$y = \frac{1}{|x|}\sum{[(x_i+2)^2 + 3]}$

</div>

You could imagine that $x$ are our parameters, and we want to optimize (either maximize or minimize) the output $y$. For this, we want to obtain the gradients $\frac{\partial{y}}{\partial{x}}$. For our example, we’ll $\hat{x} = [0,1,2]$ use  as our input.


In [2]:
x = t.arange(3, dtype=t.float32, requires_grad=True) # Only float tensor can get gradients
print("X : ", x)

X :  tensor([0., 1., 2.], requires_grad=True)


Now let’s build the computation graph step by step. You can combine multiple operations in a single line, but we will separate them here to get a better understanding of how each operation is added to the computation graph.

In [3]:
a = x + 2
b = a**2
c = b + 3
y = c.mean()
print("Y : ", y)

Y :  tensor(12.6667, grad_fn=<MeanBackward0>)


Using the statements above, we have created a computation graph that looks similar to the figure below:

<p align="center">
    <img src="./img/스크린샷 2023-01-01 오후 11.17.44.png" style="width:20%; height:20%"/>
</p>

We can perform backpropagation on the computation graph by calling the function `backward()` on the last output, which effectively calculates the gradients for each tensor that has the property `requires_grad=True`:



In [4]:
y.backward()

`x.grad` will now contain the gradient $\frac{\partial{y}}{\partial{x}}$ , and this gradient indicates how a change in $\bold{x}$  will affect output $y$  given the current input $\bold{x} = [0,1,2]$:

In [5]:
print(x.grad)

tensor([1.3333, 2.0000, 2.6667])


Hence, with the input being $\bold{x} = [0,1,2]$, our gradients are $\partial{y}/\partial{x} = [1.3333, 2.0000, 2.6667]$. The previous code cell should have printed the same result.

$x =\ 원소\ 갯수 (3개)$

$x_i =\ 원소 (0,1,2 ...)$